signingscript/sign.py

```diff
@@ -0,0 +1,434 @@
+             s.startswith("project:releng:signing:cert:")]
+    log.info("Certificate types: %s", certs)
+    if len(certs) != 1:
+        raise TaskVerificationError("Only one certificate type can be used")
+    return certs[0]
+
+
+# get_suitable_signing_servers {{{1
+def get_suitable_signing_servers(signing_servers, cert_type, signing_formats):
+    """Get the list of signing servers for given `signing_formats` and `cert_type`.
+
+    Args:
+        signing_servers (dict of lists of lists): the contents of
+            `signing_server_config`.
+        cert_type (str): the certificate type - essentially signing level,
+            separating release vs nightly vs dep.
+        signing_formats (list): the signing formats the server needs to support
+
+    Returns:
+        list of lists: the list of signing servers.
+
+    """
+    return [s for s in signing_servers[cert_type] if set(signing_formats) & set(s.formats)]
+
+
+# build_signtool_cmd {{{1
+def build_signtool_cmd(context, from_, fmt, to=None):
+    """Generate a signtool command to run.
+
+    Args:
+        context (SigningContext): the signing context
+        from_ (str): the source file to sign
+        fmt (str): the format to sign with
+        to (str, optional): the target path to sign to. If None, overwrite
+            `from_`. Defaults to None.
+
+    Returns:
+        list: the signtool command to run.
+
+    """
+    to = to or from_
+    work_dir = context.config['work_dir']
+    token = os.path.join(work_dir, "token")
+    nonce = os.path.join(work_dir, "nonce")
+    cert_type = task_cert_type(context.task)
+    ssl_cert = context.config['ssl_cert']
+    signtool = context.config['signtool']
+    if not isinstance(signtool, (list, tuple)):
+        signtool = [signtool]
+    cmd = signtool + ["-v", "-n", nonce, "-t", token, "-c", ssl_cert]
+    for s in get_suitable_signing_servers(
+        context.signing_servers, cert_type, [fmt]
+    ):
+        cmd.extend(["-H", s.server])
+    cmd.extend(["-f", fmt])
+    cmd.extend(["-o", to, from_])
+    return cmd
+
+
+# sign_file {{{1
+async def sign_file(context, from_, fmt, to=None):
+    """Send the file to signtool to be signed.
+
+    Args:
+        context (SigningContext): the signing context
+        from_ (str): the source file to sign
+        fmt (str): the format to sign with
+        to (str, optional): the target path to sign to. If None, overwrite
+            `from_`. Defaults to None.
+
+    Raises:
+        FailedSubprocess: on subprocess error while signing.
+
+    Returns:
+        str: the path to the signed file
+
+    """
+    log.info("Signing {} with {}...".format(from_, fmt))
+    cmd = build_signtool_cmd(context, from_, fmt, to=to)
+    await utils.execute_subprocess(cmd)
+    return to or from_
+
+
+# sign_gpg {{{1
+async def sign_gpg(context, from_, fmt):
+    """Create a detached signature with the gpg key.
+
+    Because this function returns a list, gpg must be the final signing format.
```

Comment:
it's mainly because we take the return value and send it to the next sign function, which expects a non-list.